core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,376 @@
+                requested = LongMath.saturatedAdd(oldRequested, n);
+            }
+
+            if (oldRequested > 0) {
+                // InputStreamSubscription is reading input stream.
+                // New requests will be handled by 'publishDownstream(HttpData)'.
+                return;
+            }
+
+            readBytes();
+        }
+
+        private void readBytes() {
+            if (requested <= 0) {
+                return;
+            }
+
+            blockingTaskExecutor.execute(() -> {
+                if (position >= end) {
+                    close(null);
+                    return;
+                }
+
+                if (position < offset) {
+                    final long skip = offset - position;
+                    final long actualSkipped;
+                    try {
+                        actualSkipped = inputStream.skip(skip);
+                    } catch (Exception e) {
+                        close(e);
+                        return;
+                    }
+
+                    if (actualSkipped < skip) {
+                        close(null);
+                        return;
+                    }
+
+                    position += skip;
+                }
+
+                final int bufferSize = Math.min(this.bufferSize, Ints.saturatedCast(end - position));
+                final int len;
+                final byte[] readBytes = new byte[bufferSize];
+                try {
+                    len = inputStream.read(readBytes);
+                } catch (Exception e) {
+                    close(e);
+                    return;
+                }
+
+                if (len == -1) {
+                    close(null);
+                    return;
+                }
+
+                final HttpData data = HttpData.wrap(readBytes, 0, len);
+                position += len;
+
+                executor.execute(() -> publishDownstream(data));
+            });
+        }
+
+        private void publishDownstream(HttpData data) {
+            if (closed) {
+                return;
+            }
+
+            if (requested != Long.MAX_VALUE) {
+                requested--;
+            }
+            downstream.onNext(data);
+            readBytes();
+        }
+
+        @Override
+        public void cancel() {
+            if (executor.inEventLoop()) {
+                cancel0();
+            } else {
+                executor.execute(this::cancel0);
+            }
+        }
+
+        private void cancel0() {
+            if (closed) {
+                return;
+            }
+            closed = true;
+
+            final CancelledSubscriptionException cause = CancelledSubscriptionException.get();
+            if (notifyCancellation) {
+                downstream.onError(cause);
+            }
+            downstream = NoopSubscriber.get();
+            completionFuture.completeExceptionally(cause);
+            maybeCloseInputStream();
+        }
+
+        private void maybeCloseInputStream() {
```

Comment:
This is not a `maybe`. It always attempts to close.

Reply:
✅ updated, `maybeCloseInputStream ` -> `closeInputStream`